packages/common/src/store/cache/syncWithReactQuery.ts

```diff
@@ -0,0 +1,147 @@
+import type { QueryClient } from '@tanstack/react-query'
+
+import { Kind } from '../../models/Kind'
+
+import type { Entry, EntriesByKind } from './types'
+
+type EntityUpdater = {
+  singleKey: string
+  listKey: string
+  idField: string
+  updateRelations?: (
+    queryClient: QueryClient,
+    id: number,
+    metadata: any
+  ) => void
+}
+
+const ENTITY_UPDATERS: Record<Kind, EntityUpdater> = {
+  [Kind.USERS]: {
+    singleKey: 'user',
+    listKey: 'users',
+    idField: 'user_id',
+    updateRelations: (queryClient, id, metadata) => {
+      // Update any tracks that might contain this user
+      queryClient.setQueriesData({ queryKey: ['track'] }, (oldData: any) => {
+        if (!oldData?.user || oldData.user.user_id !== id) return oldData
+        return {
+          ...oldData,
+          user: metadata
+        }
+      })
+
+      // Update any collections that might contain this user
+      queryClient.setQueriesData(
+        { queryKey: ['collection'] },
+        (oldData: any) => {
+          if (!oldData) return oldData
+
+          let updatedData = oldData
+
+          // Update collection if the user is the owner
+          if (oldData.user?.user_id === id) {
+            updatedData = {
+              ...oldData,
+              user: metadata
+            }
+          }
+
+          // Update collection if the user appears in any of the tracks
+          if (
+            oldData.tracks?.some((track: any) => track.user?.user_id === id)
+          ) {
+            updatedData = {
+              ...updatedData,
+              tracks: oldData.tracks.map((track: any) =>
+                track.user?.user_id === id
+                  ? {
+                      ...track,
+                      user: metadata
+                    }
+                  : track
+              )
+            }
+          }
+
+          return updatedData
+        }
+      )
+    }
+  },
+  [Kind.TRACKS]: {
+    singleKey: 'track',
+    listKey: 'tracks',
+    idField: 'track_id',
+    updateRelations: (queryClient, id, metadata) => {
+      // Update any collections that might contain this track
+      queryClient.setQueriesData(
+        { queryKey: ['collection'] },
+        (oldData: any) => {
+          if (!oldData?.tracks) return oldData
+          return {
+            ...oldData,
+            tracks: oldData.tracks.map((track: any) =>
+              track.track_id === id ? metadata : track
+            )
+          }
+        }
+      )
+    }
+  },
+  [Kind.COLLECTIONS]: {
```

Comment:
Collections doesn't need updateRelations?

Reply:
so this is just needed when an entity may be in another entity, which isn't the case for collections.